presto-main/src/main/java/io/prestosql/sql/planner/sanity/DynamicFiltersChecker.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.sanity;
+
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.Session;
+import io.prestosql.execution.warnings.WarningCollector;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.SimplePlanVisitor;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.plan.ExchangeNode;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.intersection;
+import static io.prestosql.sql.DynamicFilters.extractDynamicFilters;
+
+public class DynamicFiltersChecker
+        implements PlanSanityChecker.Checker
+{
+    @Override
+    public void validate(PlanNode plan, Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider types, WarningCollector warningCollector)
+    {
+        plan.accept(new SimplePlanVisitor<Set<String>>()
+        {
+            @Override
+            public Void visitJoin(JoinNode node, Set<String> expected)
```

Comment:
`expected` should be immutable

Reply:
To change `expected` to `ImmutableSet` we would need to change this visitor to `SimplePlanVisitor<ImmutableSet<Filter>>` and modify all the functions here. But `expected` is being mutated in all the functions here, can you please clarify why we should change this ?

Reply:
Visitor design when `context` is mutable is fragile and hard to reason about. To be more specific in this case:
1. when one visits subtree `context` will be polluted by information from other subtrees. This means that we might not remove dynamic filter producers if consumers are from sibling subtrees.
Therefore instead of using `SimplePlanVisitor` we should use `PlanVisitor`.
2. Additionally, we should remove dynamic filter consumers if there are no producers downstream or the subtree is not probe side. This requires keeping track of dynamic filters producers when visiting subtrees.

Reply:
I've made changes to use PlanVisitor and made the context immutable.